comprobo_final_project/scripts/robot_controller.py

```diff
@@ -52,28 +55,54 @@ def run(self, duration):
 
         try:
             while time.time() < end_time:
-                curr_x, curr_y, curr_w = self.robot.get_position()
-                goal_x = 0.0
-                goal_y = 0.0
+                curr_x1, curr_y1, curr_w1 = self.robot1.get_position()
+                curr_x2, curr_y2, curr_w2 = self.robot2.get_position()
+                curr_x3, curr_y3, curr_w3 = self.robot3.get_position()
+
+                # Calculate difference between robot position and other robots position
+                diff_x21 = curr_x2 - curr_x1
+                diff_y21 = curr_y2 - curr_y1
+
+                diff_x31 = curr_x3 - curr_x1
+                diff_y31 = curr_y3 - curr_y1
 
-                # Calculate difference between robot position and goal position
-                diff_x = goal_x - curr_x
-                diff_y = goal_y - curr_y
+                diff_x32 = curr_x3 - curr_x2
+                diff_y32 = curr_y3 - curr_y2
 
                 try:
                     # Calculate angle to goal and distance to goal
-                    diff_w = math.atan2(diff_y, diff_x) - curr_w
-                    diff_r = math.sqrt(diff_x**2 + diff_y**2)
+                    diff_w21 = math.atan2(diff_y21, diff_x21) - curr_w1
+                    diff_w12 = math.atan2(diff_y21, diff_x21) - curr_w2
+                    diff_r21 = math.sqrt(diff_x21**2 + diff_y21**2)
+
+                    diff_w31 = math.atan2(diff_y31, diff_x31) - curr_w1
+                    diff_w13 = math.atan2(diff_y31, diff_x31) - curr_w3
+                    diff_r31 = math.sqrt(diff_x31**2 + diff_y31**2)
+
+                    diff_w32 = math.atan2(diff_y32, diff_x32) - curr_w2
+                    diff_w23 = math.atan2(diff_y32, diff_x32) - curr_w3
+                    diff_r32 = math.sqrt(diff_x32**2 + diff_y32**2)
                 except OverflowError:
-                    print diff_x, diff_y
+                    print diff_x21, diff_y21, diff_x31, diff_y31, diff_x32, diff_y32
 
                 # Define linear and angular velocities based on genes
-                a1, b1, a2, b2 = self.genes
-                forward_rate = a1*diff_w + b1*diff_r
-                turn_rate = a2*diff_w + b2*diff_r
+                a1, b1, c1, d1, a2, b2, c2, d2, a3, b3, c3, d3 = self.genes
+
+                # calculate movements for each robot
+                forward_rate1 = a2*diff_w21 + b2*diff_r21 + a3*diff_w31 + b3*diff_r31
+                turn_rate1 = c2*diff_w21 + d2*diff_r21 + c3*diff_w31 + d3*diff_r31
+
+                forward_rate2 = a1*diff_w12 + b1*diff_r21 + a3*diff_w32 + b3*diff_r32
+                turn_rate2 = c1*diff_w12 + d1*diff_r21 + c3*diff_w32 + d3*diff_r32
+
+                forward_rate3 = a1*diff_w13 + b1*diff_r31 + a2*diff_w23 + b2*diff_r32
+                turn_rate3 = c1*diff_w13 + d1*diff_r31 + c2*diff_w23 + d2*diff_r32
```

Comment:
There is a probably a loop that could handle all this, but the form looks reasonable.